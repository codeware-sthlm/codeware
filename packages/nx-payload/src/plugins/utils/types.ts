export type PayloadPluginOptions = {
  buildTargetName?: string;
  dockerBuildTargetName?: string;
  dockerRunTargetName?: string;
  generateTargetName?: string;
  mongodbTargetName?: string;
  payloadTargetName?: string;
  postgresTargetName?: string;
  serveTargetName?: string;
  startTargetName?: string;
  stopTargetName?: string;
};

export type NormalizedOptions = Required<PayloadPluginOptions>;
