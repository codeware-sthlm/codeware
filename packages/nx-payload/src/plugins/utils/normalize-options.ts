import type { NormalizedOptions, PayloadPluginOptions } from './types';

export const normalizeOptions = (
  options?: PayloadPluginOptions
): NormalizedOptions => ({
  buildTargetName: options?.buildTargetName ?? 'build',
  generateTargetName: options?.generateTargetName ?? 'gen',
  payloadTargetName: options?.payloadTargetName ?? 'payload',
  serveTargetName: options?.serveTargetName ?? 'serve',
  dxDockerBuildTargetName:
    options?.dxDockerBuildTargetName ?? 'dx:docker-build',
  dxDockerRunTargetName: options?.dxDockerRunTargetName ?? 'dx:docker-run',
  dxMongodbTargetName: options?.dxMongodbTargetName ?? 'dx:mongodb',
  dxPostgresTargetName: options?.dxPostgresTargetName ?? 'dx:postgres',
  dxStartTargetName: options?.dxStartTargetName ?? 'dx:start',
  dxStopTargetName: options?.dxStopTargetName ?? 'dx:stop'
});
