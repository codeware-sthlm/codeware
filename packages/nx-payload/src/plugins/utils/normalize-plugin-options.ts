export type PayloadPluginOptions = {
  generateTargetName?: string;
  payloadTargetName?: string;
  payloadGraphqlTargetName?: string;
  dxMongodbTargetName?: string;
  dxPostgresTargetName?: string;
  dxStartTargetName?: string;
  dxStopTargetName?: string;
};

type NormalizedOptions = Required<PayloadPluginOptions>;

/**
 * Apply default values to plugin options that are not set.
 */
export const normalizePluginOptions = (
  options?: PayloadPluginOptions
): NormalizedOptions => ({
  generateTargetName: options?.generateTargetName ?? 'gen',
  payloadTargetName: options?.payloadTargetName ?? 'payload',
  payloadGraphqlTargetName:
    options?.payloadGraphqlTargetName ?? 'payload-graphql',
  dxMongodbTargetName: options?.dxMongodbTargetName ?? 'dx:mongodb',
  dxPostgresTargetName: options?.dxPostgresTargetName ?? 'dx:postgres',
  dxStartTargetName: options?.dxStartTargetName ?? 'dx:start',
  dxStopTargetName: options?.dxStopTargetName ?? 'dx:stop'
});
