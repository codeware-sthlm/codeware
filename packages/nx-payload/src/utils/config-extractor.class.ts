import ts from 'typescript';

import type { ConfigPropertyPath } from './extract-payload-config';

// Helper type to allow any once
// eslint-disable-next-line @typescript-eslint/no-explicit-any
type ANY = any;

/**
 * Extract the Payload config object from the provided absolute path.
 *
 * The extraction is done by parsing the TypeScript source file,
 * which comes with some limitations by design:
 *
 * - The config object must be exported using `export default buildConfig()`
 * - Functions are replaced with their string representation in the result
 * - The config object is assumed to be a valid Payload config when returned,
 *   due to the file being type-safe by design
 *
 * @todo Better documentation
 */
export class ConfigExtractor {
  private variables: Map<string, ANY>;
  private checker: ts.TypeChecker;
  private configObject: ANY | null;
  private targetProperties: Set<string> | null;
  private processedPaths: Set<string>;

  constructor(
    private sourceFile: ts.SourceFile,
    checker: ts.TypeChecker,
    properties?: ConfigPropertyPath | Array<ConfigPropertyPath>
  ) {
    this.variables = new Map();
    this.checker = checker;
    this.configObject = null;
    this.processedPaths = new Set();
    this.targetProperties = this.normalizeProperties(properties);
  }

  /**
   * Extracts the Payload config object from the source file.
   *
   * This method traverses the AST of the source file to find the export
   * assignment of the buildConfig function and extracts the config object.
   *
   * @returns The extracted config object or null if not found.
   */
  extract(): ANY | null {
    this.collectVariables();
    const findBuildConfig = (node: ts.Node): void => {
      if (ts.isExportAssignment(node)) {
        const expression = node.expression;
        if (
          ts.isCallExpression(expression) &&
          expression.expression.getText() === 'buildConfig'
        ) {
          const [configArg] = expression.arguments;
          if (configArg) {
            if (ts.isObjectLiteralExpression(configArg)) {
              this.configObject = this.parseObjectLiteral(configArg);
            } else if (ts.isIdentifier(configArg)) {
              const configVar = configArg.getText();
              const config = this.variables.get(configVar);
              if (config) {
                this.configObject = config;
              }
            }
          }
        }
      }
      ts.forEachChild(node, findBuildConfig);
    };

    findBuildConfig(this.sourceFile);
    return this.configObject;
  }

  /**
   * Gets the properties that are missing from the processed paths.
   *
   * This method compares the target properties with the processed paths
   * to identify which properties have not been processed.
   *
   * @returns An array of strings representing the missing properties.
   */
  getMissingProperties(): Array<string> {
    return Array.from(this.targetProperties ?? []).filter(
      (prop) => !this.processedPaths.has(prop)
    );
  }

  /**
   * Normalizes the provided properties into a Set of strings.
   *
   * This method ensures that the properties are stored in a consistent format
   * for further processing.
   *
   * @param properties - A single property or an array of properties to normalize.
   * @returns A Set of normalized property paths or null if no properties are provided.
   */
  private normalizeProperties(
    properties?: ConfigPropertyPath | Array<ConfigPropertyPath>
  ): Set<string> | null {
    if (!properties) {
      return null;
    }

    const props = Array.isArray(properties) ? properties : [properties];

    return new Set(props);
  }

  /**
   * Determines whether a property should be processed based on the target properties.
   *
   * This method checks if the property path matches any of the target properties,
   * allowing for root paths, exact matches, and wildcard matches (e.g., 'admin.*').
   * It also tracks processed paths to avoid reprocessing.
   *
   * @param propertyPath - An array of strings representing the path of the property.
   * @returns A boolean indicating if the property should be processed.
   */
  private shouldProcessProperty(propertyPath: Array<string>): boolean {
    if (!this.targetProperties) {
      return true;
    }

    const fullPath = propertyPath.join('.');
    this.processedPaths.add(fullPath);

    return Array.from(this.targetProperties).some((targetPath) => {
      // For root path, always process
      if (fullPath === '') return true;

      // Exact match
      if (targetPath === fullPath) return true;

      // When requesting 'admin', we need to process 'admin.*'
      if (targetPath === propertyPath[0]) return true;

      // When requesting 'admin.something', we need to process 'admin'
      if (fullPath === targetPath.split('.')[0]) return true;

      // When requesting nested property, process all parts of the path
      return (
        targetPath.startsWith(fullPath + '.') ||
        fullPath.startsWith(targetPath + '.')
      );
    });
  }

  /**
   * Collects variable declarations from the source file.
   *
   * This method traverses the AST to find variable statements and stores
   * their initial values in the variables map for later use.
   */
  private collectVariables(): void {
    const visit = (node: ts.Node): void => {
      if (ts.isVariableStatement(node)) {
        for (const declaration of node.declarationList.declarations) {
          if (declaration.initializer) {
            const name = declaration.name.getText();

            if (ts.isObjectLiteralExpression(declaration.initializer)) {
              const obj: Record<string, ANY> = {};

              for (const prop of declaration.initializer.properties) {
                if (ts.isPropertyAssignment(prop)) {
                  const propName = prop.name.getText();
                  const value = this.parseNode(prop.initializer);
                  obj[propName] = value;
                }
              }

              this.variables.set(name, obj);
            } else {
              const value = this.parseNode(declaration.initializer);
              this.variables.set(name, value);
            }
          }
        }
      }
      ts.forEachChild(node, visit);
    };

    visit(this.sourceFile);
  }

  /**
   * Resolves the value of a variable based on its identifier or property access.
   *
   * This method checks if the variable is defined and retrieves its value
   * from the variables map or resolves it through property access.
   *
   * @param node - The AST node representing the variable or property access.
   * @param propertyPath - An array of strings representing the path of the property.
   * @returns The resolved value or undefined if not found.
   */
  private resolveVariableValue(
    node: ts.Node,
    propertyPath: Array<string> = []
  ): ANY {
    if (!this.shouldProcessProperty(propertyPath)) {
      return undefined;
    }

    if (ts.isIdentifier(node)) {
      return this.variables.get(node.getText());
    }

    if (ts.isPropertyAccessExpression(node)) {
      const baseValue = this.resolveVariableValue(
        node.expression,
        propertyPath
      );
      if (baseValue && typeof baseValue === 'object') {
        return baseValue[node.name.getText()];
      }
    }

    return undefined;
  }

  /**
   * Parses a given AST node to extract its value based on its type.
   *
   * This method handles different node types (e.g., object literals, arrays,
   * literals) and delegates to other methods for specific parsing logic.
   *
   * @param node - The AST node to parse.
   * @param propertyPath - An array of strings representing the path of the property.
   * @returns The parsed value or undefined if not processed.
   */
  private parseNode(node: ts.Node, propertyPath: Array<string> = []): ANY {
    if (!this.shouldProcessProperty(propertyPath)) {
      return undefined;
    }

    switch (node.kind) {
      case ts.SyntaxKind.ObjectLiteralExpression:
        return this.parseObjectLiteral(
          node as ts.ObjectLiteralExpression,
          propertyPath
        );
      case ts.SyntaxKind.ArrayLiteralExpression:
        return (node as ts.ArrayLiteralExpression).elements.map(
          (element, index) =>
            this.parseNode(element, [...propertyPath, index.toString()])
        );
      case ts.SyntaxKind.StringLiteral:
        return (node as ts.StringLiteral).text;
      case ts.SyntaxKind.NumericLiteral:
        return Number((node as ts.NumericLiteral).text);
      case ts.SyntaxKind.TrueKeyword:
        return true;
      case ts.SyntaxKind.FalseKeyword:
        return false;
      case ts.SyntaxKind.CallExpression:
        return node.getText();
      case ts.SyntaxKind.PropertyAccessExpression: {
        const resolved = this.resolveVariableValue(node, propertyPath);
        return resolved !== undefined ? resolved : node.getText();
      }
      case ts.SyntaxKind.Identifier: {
        const resolved = this.resolveVariableValue(node, propertyPath);
        return resolved !== undefined ? resolved : node.getText();
      }
      default:
        return node.getText();
    }
  }

  /**
   * Parses an object literal expression to extract its properties and values.
   *
   * This method iterates over the properties of the object literal and
   * processes each property based on its type and the current property path.
   *
   * @param node - The object literal expression to parse.
   * @param propertyPath - An array of strings representing the path of the property.
   * @returns A record of property names and their corresponding values.
   */
  private parseObjectLiteral(
    node: ts.ObjectLiteralExpression,
    propertyPath: Array<string> = []
  ): Record<string, ANY> {
    const result: Record<string, ANY> = {};

    for (const property of node.properties) {
      if (ts.isPropertyAssignment(property)) {
        const propName = property.name.getText();
        const newPath = [...propertyPath, propName];

        if (this.shouldProcessProperty(newPath)) {
          const value = this.parseNode(property.initializer, newPath);
          if (value !== undefined) {
            result[propName] = value;
          }
        }
      } else if (ts.isShorthandPropertyAssignment(property)) {
        const name = property.name.getText();
        const newPath = [...propertyPath, name];

        if (this.shouldProcessProperty(newPath)) {
          const value = this.resolveVariableValue(property.name, newPath);
          if (value !== undefined) {
            result[name] = value;
          } else {
            result[name] = property.name.getText();
          }
        }
      }
    }

    return result;
  }
}
