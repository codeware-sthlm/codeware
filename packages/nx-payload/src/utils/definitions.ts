/**
 * Payload plugin targets.
 *
 * These targets are dedicated to Payload actions and developer experience.
 *
 * **Note:** `generate` is aliased to `gen` to avoid conflict with native `nx generate`.
 */
export const payloadTargets = [
  'gen',
  'payload',
  'payload-graphql',
  'dx:mongodb',
  'dx:postgres',
  'dx:start',
  'dx:stop'
] as const;

/**
 * Payload plugin targets.
 *
 * These targets are dedicated to Payload actions and developer experience.
 *
 * **Note:** `generate` is aliased to `gen` to avoid conflict with native `nx generate`.
 */
export type PayloadTarget = (typeof payloadTargets)[number];
