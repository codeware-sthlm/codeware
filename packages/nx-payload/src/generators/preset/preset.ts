import {
  type GeneratorCallback,
  type Tree,
  formatFiles,
  readNxJson,
  runTasksInSerial
} from '@nx/devkit';

import { isPluginInferenceEnabled } from '../../utils/is-plugin-inference-enabled';
import { applicationGenerator } from '../application/application';

import { normalizeOptions } from './libs/normalize-options';
import type { PresetGeneratorSchema } from './schema';

/**
 * Payload application preset when a workspace is created with Nx CLI.
 *
 * Defined as factory function in `generators.json` and hence invoked by Nx CLI.
 */
export async function presetGenerator(
  host: Tree,
  schema: PresetGeneratorSchema
) {
  const tasks: Array<GeneratorCallback> = [];

  const nxJson = readNxJson(host);
  const options = normalizeOptions(host, schema);

  // Generate application
  const appGenTask = await applicationGenerator(host, {
    ...options,
    addPlugin: isPluginInferenceEnabled(nxJson),
    skipFormat: true
  });
  tasks.push(appGenTask);

  host.delete('libs');

  if (!options.skipFormat) {
    await formatFiles(host);
  }

  return runTasksInSerial(...tasks);
}
