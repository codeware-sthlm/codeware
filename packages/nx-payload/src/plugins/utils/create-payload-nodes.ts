import { dirname, join, relative } from 'path/posix';

import {
  type CreateNodesContext,
  type CreateNodesResult,
  type ProjectConfiguration,
  type TargetConfiguration,
  detectPackageManager,
  normalizePath,
  readJsonFile
} from '@nx/devkit';
import { calculateHashForCreateNodes } from '@nx/devkit/src/utils/calculate-hash-for-create-nodes';
import { getLockFileName } from '@nx/js';

import { createPayloadTargets } from '../../utils/create-payload-targets';
import { extractPayloadConfig } from '../../utils/extract-payload-config';
import { findUpFs } from '../../utils/find-up-fs';

import {
  type PayloadPluginOptions,
  normalizePluginOptions
} from './normalize-plugin-options';

/**
 * Create Payload nodes for a project
 *
 * @param configFilePath - Path to the Payload config file provided by `createNodesFromFiles`
 * @param context - Create nodes context
 * @param options - Plugin options
 * @param targetsCache - Targets cache
 */
export const createPayloadNodes = async (
  configFilePath: string,
  context: CreateNodesContext,
  options: PayloadPluginOptions | undefined,
  targetsCache: Record<
    string,
    Record<string, TargetConfiguration<PayloadPluginOptions>>
  >
): Promise<CreateNodesResult> => {
  // Convert to POSIX on Windows (workspacePath is os formatted)
  const workspaceRoot = normalizePath(context.workspaceRoot);

  // configFilePath is POSIX formatted
  const configRoot = dirname(configFilePath);
  const fullConfigRoot = join(workspaceRoot, configRoot);

  // Payload config file can be located anywhere, though it's generated initially in the 'src' folder.
  // Find project root by looking for project.json in a parent directory.
  const projectJsonPath = await findUpFs('project.json', {
    startPath: fullConfigRoot,
    stopAtPath: workspaceRoot
  });

  // Do not create targets if project.json isn't found.
  // Plugin limitation: Nx native plugins also allow package.json to be used as a project file.
  if (!projectJsonPath) {
    return {};
  }

  // Create path for project root, e.g. 'apps/my-app'
  const projectRoot = relative(workspaceRoot, dirname(projectJsonPath));

  const { name: projectName } =
    readJsonFile<ProjectConfiguration>(projectJsonPath);

  // Get current GraphQL config state
  const result = extractPayloadConfig(configRoot, 'graphQL.disable');
  if (!result.config) {
    // Don't throw to prevent messing up the targets
    console.error(result.error);
    console.warn(
      `Could not extract Payload config from ${configFilePath}, GraphQL state is not known`
    );
  }
  const isGraphQLDisabled = result?.config?.graphQL?.disable ?? true;

  const hash = await calculateHashForCreateNodes(
    projectRoot,
    normalizePluginOptions(options),
    context,
    [
      getLockFileName(detectPackageManager(workspaceRoot)),
      String(isGraphQLDisabled)
    ]
  );

  // Get payload targets to be inferred for the project
  targetsCache[hash] ??= createPayloadTargets({
    isGraphQLDisabled,
    projectName: String(projectName), // Should have a value by design?
    projectRoot
  });

  return {
    projects: {
      [projectRoot]: {
        root: projectRoot,
        targets: targetsCache[hash]
      }
    }
  };
};
