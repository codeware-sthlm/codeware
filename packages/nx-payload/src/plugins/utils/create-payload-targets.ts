import type {
  CreateNodesContext,
  ProjectConfiguration,
  TargetConfiguration
} from '@nx/devkit';
import { getNamedInputs } from '@nx/devkit/src/utils/get-named-inputs';
import type { RunCommandsOptions } from 'nx/src/executors/run-commands/run-commands.impl';

import type { BuildExecutorSchema } from '../../executors/build/schema';
import { metadata } from '../../utils/definitions';

import type { NormalizedOptions } from './types';

export const createPayloadTargets = async (
  projectRoot: string,
  projectConfig: ProjectConfiguration,
  options: NormalizedOptions,
  context: CreateNodesContext
): Promise<Record<string, TargetConfiguration>> => {
  const namedInputs = getNamedInputs(projectRoot, context);

  const targets: Record<
    string,
    TargetConfiguration<BuildExecutorSchema | Partial<RunCommandsOptions>>
  > = {};

  // Add build target
  targets[options.buildTargetName] = {
    metadata: metadata.build,
    executor: '@cdwr/nx-payload:build',
    inputs: [
      'default',
      Object.hasOwn(namedInputs, 'production') ? '^production' : '^default'
    ],
    outputs: ['{options.outputPath}'],
    options: {
      main: '{projectRoot}/src/main.ts',
      tsConfig: '{projectRoot}/tsconfig.app.json',
      outputPath: '{workspaceRoot}/dist/{projectRoot}',
      outputFileName: 'src/main.js'
    },
    cache: true
  };

  // Add serve target
  targets[options.serveTargetName] = {
    metadata: metadata.serve,
    executor: '@nx/js:node',
    options: {
      buildTarget: `${projectConfig.name}:${options.buildTargetName}`,
      // Required for file changes to be detected and build target to be re-run
      runBuildTargetDependencies: true,
      watch: true
    }
  };

  // Add generate target
  targets[options.generateTargetName] = {
    metadata: metadata.gen,
    executor: 'nx:run-commands',
    options: {
      commands: [
        'npx payload generate:types',
        'npx payload generate:graphQLSchema'
      ],
      env: {
        PAYLOAD_CONFIG_PATH: '{projectRoot}/src/payload.config.ts'
      },
      parallel: false
    }
  };

  // Add payload target
  targets[options.payloadTargetName] = {
    metadata: metadata.payload,
    executor: 'nx:run-commands',
    options: {
      command: 'npx payload',
      forwardAllArgs: true,
      env: {
        PAYLOAD_CONFIG_PATH: '{projectRoot}/src/payload.config.ts'
      }
    }
  };

  // Add mongodb target
  targets[options.dxMongodbTargetName] = {
    metadata: metadata['dx:mongodb'],
    command: `docker ps -q -f name=mongodb-${projectConfig.name} | grep . && echo '[Running] mongodb is already started' || docker run --name mongodb-${projectConfig.name} --rm -d -p 27017:27017 mongo`
  };

  // Add postgres target
  targets[options.dxPostgresTargetName] = {
    metadata: metadata['dx:postgres'],
    executor: 'nx:run-commands',
    options: {
      command: `docker ps -q -f name=postgres-${projectConfig.name} | grep . && echo '[Running] PostgreSQL init process complete' || docker run --name postgres-${projectConfig.name} --rm --env-file ${projectRoot}/.env.local -p 5432:5432 postgres`,
      readyWhen: 'PostgreSQL init process complete'
    }
  };

  // Add start target
  targets[options.dxStartTargetName] = {
    metadata: metadata['dx:start'],
    command: `docker compose -f ${projectRoot}/docker-compose.yml up -d`
  };

  // Add stop target
  targets[options.dxStopTargetName] = {
    metadata: metadata['dx:stop'],
    command: `docker compose -f ${projectRoot}/docker-compose.yml down`
  };

  // Add docker-build target
  targets[options.dxDockerBuildTargetName] = {
    metadata: metadata['dx:docker-build'],
    command: `docker build -f ${projectRoot}/Dockerfile -t ${projectConfig.name} .`
  };

  // Add docker-run target
  targets[options.dxDockerRunTargetName] = {
    metadata: metadata['dx:docker-run'],
    command: `docker run --name ${projectConfig.name} --rm --env-file ${projectRoot}/.env.local -d -p 3000:3000 ${projectConfig.name}`
  };

  return targets;
};
