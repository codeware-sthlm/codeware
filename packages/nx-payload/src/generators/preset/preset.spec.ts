import { type Tree, readProjectConfiguration } from '@nx/devkit';
import { createTreeWithEmptyWorkspace } from '@nx/devkit/testing';

import { payloadTargets } from '../../utils/definitions';

import { presetGenerator } from './preset';
import type { PresetGeneratorSchema } from './schema';

describe.skip('preset generator', () => {
  let tree: Tree;

  const options: PresetGeneratorSchema = {
    name: 'test',
    payloadAppName: 'test-app',
    payloadAppDirectory: 'app-dir/test-app',
    addPlugin: true,
    skipFormat: true
  };

  console.log = jest.fn();
  console.warn = jest.fn();

  jest.setTimeout(10_000);

  let nxDaemon: string;
  beforeAll(() => {
    nxDaemon = process.env['NX_DAEMON'];
    process.env['NX_DAEMON'] = 'false';
  });

  afterAll(() => {
    process.env['NX_DAEMON'] = nxDaemon;
  });

  beforeEach(() => {
    jest.clearAllMocks();
    tree = createTreeWithEmptyWorkspace({ layout: 'apps-libs' });
  });

  it('should generate project config without payload targets', async () => {
    await presetGenerator(tree, options);

    const { targets, ...config } = readProjectConfiguration(tree, 'test-app');

    for (const target of payloadTargets) {
      expect(targets[target]).toBeUndefined();
    }

    expect(config).toMatchSnapshot();
  });

  it('should use workspace `name` when `appName` is not provided', async () => {
    await presetGenerator(tree, {
      ...options,
      name: 'workspace-name',
      payloadAppName: '',
      payloadAppDirectory: ''
    });

    const { name } = readProjectConfiguration(tree, 'workspace-name');
    expect(name).toBe('workspace-name');
  });

  it('should set "apps" as the default app base path', async () => {
    await presetGenerator(tree, {
      ...options,
      name: 'test',
      payloadAppName: 'test-app',
      payloadAppDirectory: ''
    });

    const { name, sourceRoot } = readProjectConfiguration(tree, 'test-app');
    expect(name).toBe('test-app');
    expect(sourceRoot).toBe('apps/test-app');
  });

  it('should delete "libs" folder', async () => {
    await presetGenerator(tree, options);

    expect(tree.children('').includes('apps')).toBeTruthy();
    expect(tree.children('').includes('libs')).toBeFalsy();
  });
});
