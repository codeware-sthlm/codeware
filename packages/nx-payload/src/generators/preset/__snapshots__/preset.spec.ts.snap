// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`preset generator should generate project config without payload targets 1`] = `
{
  "$schema": "../../node_modules/nx/schemas/project-schema.json",
  "name": "test-app",
  "projectType": "application",
  "root": "app-dir/test-app",
  "sourceRoot": "app-dir/test-app",
  "tags": [],
}
`;
